--- pylibftdi/pylibftdi/driver.py	2019-10-19 00:06:46.346310880 -0700
+++ pylibftdi/pylibftdi/driver.py	2019-10-19 00:04:18.956315155 -0700
@@ -134,7 +134,7 @@
         primarily for diagnostic purposes.
         """
         if self._libusb_dll is None:
-            self._libusb_dll = self._load_library('libusb')
+            self._libusb_dll = find_library('libusb1')
             self._libusb_dll.libusb_get_version.restype = POINTER(libusb_version_struct)

         return self._libusb_dll
@@ -156,7 +156,7 @@
         This is the main interface to FTDI functionality.
         """
         if self._fdll is None:
-            self._fdll = self._load_library('libftdi')
+            self._fdll = find_library('libftdi1')
             # most args/return types are fine with the implicit
             # int/void* which ctypes uses, but some need setting here
             self._fdll.ftdi_get_error_string.restype = c_char_p
