From 20043beb6bf4b002da1f79a0b102c6607992fbbd Mon Sep 17 00:00:00 2001
From: Matt Huszagh <huszaghmatt@gmail.com>
Date: Tue, 3 Dec 2019 13:03:38 -0800
Subject: [PATCH] add sympy antlr dep

---
 .../python-modules/sympy/default.nix           | 18 +++++++++++++++++-
 1 file changed, 17 insertions(+), 1 deletion(-)

diff --git a/pkgs/development/python-modules/sympy/default.nix b/pkgs/development/python-modules/sympy/default.nix
index c09267662c1..9b8d154f4c5 100644
--- a/pkgs/development/python-modules/sympy/default.nix
+++ b/pkgs/development/python-modules/sympy/default.nix
@@ -1,9 +1,13 @@
 { lib
 , buildPythonPackage
+, isPy3k
+, isPy27
 , fetchPypi
 , fetchpatch
 , glibcLocales
 , mpmath
+, antlr4-python3-runtime
+, antlr4-python2-runtime
 }:
 
 buildPythonPackage rec {
@@ -17,7 +21,19 @@ buildPythonPackage rec {
 
   checkInputs = [ glibcLocales ];
 
-  propagatedBuildInputs = [ mpmath ];
+  propagatedBuildInputs = [
+    mpmath
+  ] ++ lib.optionals isPy3k [
+    antlr4-python3-runtime
+  ];
+
+  # propagatedBuildInputs = [
+  #   mpmath
+  # ] ++ lib.optionals isPy3k [
+  #   antlr4-python3-runtime
+  # ] ++ lib.optionals isPy27 [
+  #   antlr4-python2-runtime
+  # ];
 
   # tests take ~1h
   doCheck = false;
-- 
2.24.0

