--- pylibftdi/pylibftdi/examples/bit_server.py	2019-10-18 23:49:44.119822248 -0700
+++ pylibftdi/pylibftdi/examples/bit_server.py	2019-10-18 23:43:48.503745669 -0700
@@ -12,9 +12,9 @@
 import threading
 import time
 import webbrowser
-from BaseHTTPServer import HTTPServer, BaseHTTPRequestHandler
-from cStringIO import StringIO
-from SocketServer import ThreadingMixIn
+from http.server import HTTPServer, BaseHTTPRequestHandler
+from io import StringIO
+from socketserver import ThreadingMixIn
 from pylibftdi import BitBangDevice

 HTTP_PORT = 8008
--- pylibftdi/pylibftdi/examples/serial_transfer.py	2019-10-18 23:49:55.342642348 -0700
+++ pylibftdi/pylibftdi/examples/serial_transfer.py	2019-10-18 23:45:26.864055962 -0700
@@ -35,7 +35,7 @@
 import threading
 from itertools import islice

-from pylibftdi import Device, ReadTimeout
+from pylibftdi import Device


 class RandomStream(object):
